### ISSUE TITLE
Pass variables to the view

### ISSUE THREAD

**Issue by franciscop** [yan-foto/electron-jade#9]
While I understand that you can pass variables on startup like this:

```js
var j = require('electron-jade')({pretty: true}, locals);
```

Where `locals` are the variables, is there any way to pass them later on? I have a `request` call and I need to load that like this:

```js
var app = require('app');
var request = require('request');
var j = require('electron-jade')({pretty: true}, locals);
//...
app.on('ready', function() {
    request('blabla', function(err, status, body){

      // Here I want to pass some variables from "body":
      window.loadUrl('file://' + __dirname + '/index.jade');
    });
});
```

Writing the require before `window.loadUrl` doesn't work:

```js
var app = require('app');
var request = require('request');
var j = require('electron-jade')({pretty: true}, locals);
//...
app.on('ready', function() {
    request('blabla', function(err, status, body){

      // Doesn't work:
      var j = require('electron-jade')({pretty: true}, { body: body });
      window.loadUrl('file://' + __dirname + '/index.jade');
    });
});
```

**Comment by franciscop** [comment]
A note, within the view I can see the variable `process`, since this works as in an example. What else?

The example:

```jade
p.
  We are using io.js #{ process.version }
  and Electron #{ process.versions['electron'] }
```

**Comment by yan-foto** [comment]
**tl;dr**

ATM the API doesn't support that, but I will try to implement it ASAP. A pull request is also welcome :+1: 

---

That is some interesting use case! So lets get to the bottom of it:

> is there any way to pass them later on?

Not in the way the API is implemented.

Since objects passed to functions are [passed by reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Functions), as a workaround for now you could keep a reference to your `locals` and update it for each request. However, this is not something that I find *clean*!

> within the view I can see the variable `process`

the [`process`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Functions) object is available in the global context of the Javascript runtime within electron and that is why you can access that.

Using objects in the global context is also not a solution that I would suggest.

**Comment by franciscop** [comment]
The problem is that I have no idea how I could do it, I come from node.js and I could handle middleware and javascript, but in Electron some things seem more "close to the metal". For instance, the `window.loadUrl` source code comes from webContents, and the source for that appears to be C.

Right now what I'm doing is making a call from the *browser* javascript like this:

```js
window.$ = window.jQuery = require('./jquery.js');
var ipc = require('ipc');
ipc.on('asynchronous-reply', function(arg) {
  ...
});
ipc.send('asynchronous-message', 'init');
```

And answering from the server side like this:

```js
var ipc = require('ipc');
ipc.on('asynchronous-message', function(event, arg) {
  request({ url: config.api, json: true }, function(err, status, body){
    event.sender.send('asynchronous-reply', JSON.stringify(body));
  });
});
```

As things are local this takes no time and allows me to send data from back to front in an asynchronous manner. About `process`, is it possible to inject a variable? Because maybe we could use it to inject it like this:

var app = require('app');
var request = require('request');
var j = require('electron-jade')({pretty: true}, locals);
//...
app.on('ready', function() {
    request('blabla', function(err, status, body){

      window.process.locals = body;
      window.loadUrl('file://' + __dirname + '/index.jade');
    });
});

**Comment by yan-foto** [comment]
I think you are mixing a number of concepts together. Electron uses io.js as its javascript runtime and a modified version of chrome for rendering the front end and both are written in C/C++.
Middleware as a design pattern will not help us solving this problem (at least not in the way the API is implemented).
Until the new implementation is ready, You should take a look out at how JS handles call be reference. I suppose it's gonna be the easiest way.
On Oct 6, 2015 10:15, Francisco Presencia <notifications@github.com> wrote:The problem is that I have no idea how I could do it, I come from node.js and I could handle middleware and javascript, but in Electron some things seem more "close to the metal". For instance, the window.loadUrl source code comes from webContents, and the source for that appears to be C.

Right now what I'm doing is making a call from the browser javascript like this:

window.$ = window.jQuery = require('./jquery.js');
var ipc = require('ipc');
ipc.on('asynchronous-reply', function(arg) {
  ...
});
ipc.send('asynchronous-message', 'init');

And answering from the server side like this:

var ipc = require('ipc');
ipc.on('asynchronous-message', function(event, arg) {
  request({ url: config.api, json: true }, function(err, status, body){
    event.sender.send('asynchronous-reply', JSON.stringify(body));
  });
});

As things are local this takes no time and allows me to send data from back to front in an asynchronous manner. About process, is it possible to inject a variable? Because maybe we could use it to inject it like this:

var app = require('app');
var request = require('request');
var j = require('electron-jade')({pretty: true}, locals);
//...
app.on('ready', function() {
    request('blabla', function(err, status, body){

  window.process.locals = body;
  window.loadUrl('file://' + __dirname + '/index.jade');
});


});

**Comment by franciscop** [comment]
Thanks for the reply, this might also help:

> [loadHtml instead of loadUrl in atom-shell](https://github.com/atom/electron/issues/968)

**Comment by franciscop** [comment]
As an alternative to interjection, I've created a method to directly render it from window similar to the one I linked before. I'll try to get it accepted in `browser-window` or write an npm module. It's basically this:

```js
var BrowserWindow = require('browser-window');

// Extend BrowserWindow to allow jade parsing (maybe better loadJade())
BrowserWindow.prototype.jade = require('../window-jade');

window = new BrowserWindow({ width: 400, height: 500 });

// Several ways to include it, absolute or relative, and with extension or without:
// - `/home/.../project/print` or `/home/.../project/print.jade`
// - `print` or `print.jade`
window.jade('print', { a: "Some text" });
```

This is the file `window-jade.js`:

```js
// Load from a jade file and pass some variables to it
var jade = require('jade');

module.exports = function(file, locals){
  file = file.split('.').pop() == 'jade' ? file : file + '.jade';
  file = process.cwd() + '/' + file;
  var html = jade.renderFile(file, locals);
  window.loadUrl('data:text/html,' + encodeURIComponent(html));
};
```

**Comment by franciscop** [comment]
The previous method works great when you don't need any other assets. Otherwise it gets them all screwed, since it considers the string itself as the baseUrl so it cannot load any asset. If you try to include them with jade it works, but then you cannot load things with `require()` like `var ipc = require('ipc');`. So basically you're screwed by using anything complex with the previous file I added.

We could go down the path of generating the html files as temporary files and then doing `window.loadUrl` with that temporary file. But maybe that's just too complex.

**Comment by yan-foto** [comment]
I'm taking a week off. As soon as I'm back I'll a take a look at. We're gonna find a way, no worries!
On Oct 7, 2015 09:30, Francisco Presencia <notifications@github.com> wrote:The previous method works great when you don't need any other assets. Otherwise it gets them all screwed, since it considers the string itself as the baseUrl so it cannot load any asset. If you try to include them with jade it works, but then you cannot load things with require() like var ipc = require('ipc');. So basically you're screwed by using anything complex with the previous file I added.

We could go down the path of generating the html files as temporary files and then doing window.loadUrl with that temporary file. But maybe that's just too complex.

**Comment by tjanpool** [comment]
I know this is a closed topic, however I was using this topic to load context to my pages. Doing so I discovered that I had a problem with the keyword "window" in window-jade.js. It is better to replace "window"  with the keyword "this". By doing so the application does not depend any more on the name you give when calling <a-name> = new BrowserWindow(....). The <a-name> has to be "window" if you use the current example script
